drivers/gpio/gpio_nrfx.c

```diff
@@ -0,0 +1,458 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gpio.h>
+#include <hal/nrf_gpio.h>
+#include <hal/nrf_gpiote.h>
+
+#include "gpio_utils.h"
+
+struct gpio_nrfx_data {
+	sys_slist_t callbacks;
+
+	/* Mask holding information about which pins have been configured to
+	 * trigger interrupts using gpio_nrfx_config function.
+	 */
+	u32_t pin_int_en;
+
+	/* Mask holding information about which pins have enabled callbacks
+	 * using gpio_nrfx_enable_callback function.
+	 */
+	u32_t int_en;
+
+	u32_t active_level;
+	u32_t trig_edge;
+	u32_t double_edge;
+	u32_t inverted;
+};
+
+struct gpio_nrfx_cfg {
+	NRF_GPIO_Type *port;
+	u8_t port_num;
+};
+
+static inline struct gpio_nrfx_data *get_port_data(struct device *port)
+{
+	return port->driver_data;
+}
+
+static inline const struct gpio_nrfx_cfg *get_port_cfg(struct device *port)
+{
+	return port->config->config_info;
+}
+
+static int gpiote_channel_alloc(u32_t abs_pin, nrf_gpiote_polarity_t polarity)
+{
+	for (u8_t channel = 0; channel < GPIOTE_CH_NUM; ++channel) {
+		/* @TODO Add GPIOTE HAL function for checking
+		 * if channel is enabled
+		 */
+		if ((NRF_GPIOTE->CONFIG[channel] & GPIOTE_CONFIG_MODE_Msk)
+		    == GPIOTE_CONFIG_MODE_Disabled) {
+			nrf_gpiote_event_configure(channel, abs_pin, polarity);
+			nrf_gpiote_event_enable(channel);
+			nrf_gpiote_int_enable(BIT(channel));
+			return channel;
+		}
+	}
+
+	return -ENODEV;
+}
+
+static void gpiote_channel_free(u32_t abs_pin)
+{
+	u32_t intenset = nrf_gpiote_int_is_enabled(NRF_GPIOTE_INT_IN_MASK);
+
+	for (size_t i = 0; i < GPIOTE_CH_NUM; i++) {
+		if ((nrf_gpiote_event_pin_get(i) == abs_pin)
+		    && (intenset & BIT(i))) {
+			nrf_gpiote_event_disable(i);
+			nrf_gpiote_int_disable(BIT(i));
+			return;
+		}
+	}
+}
+
+static int gpiote_pin_int_cfg(struct device *port, u32_t pin)
+{
+	struct gpio_nrfx_data *data = get_port_data(port);
+	const struct gpio_nrfx_cfg *cfg = get_port_cfg(port);
+	u32_t abs_pin = NRF_GPIO_PIN_MAP(cfg->port_num, pin);
+	int res = 0;
+
+	gpiote_channel_free(abs_pin);
+	nrf_gpio_cfg_sense_set(abs_pin, NRF_GPIO_PIN_NOSENSE);
+
+	/* Pins trigger interrupts only if pin has been configured to do so
+	 * and callback has been enabled for that pin.
+	 */
+	if ((data->pin_int_en & BIT(pin)) && (data->int_en & BIT(pin))) {
+		if (data->trig_edge & BIT(pin)) {
+		/* For edge triggering we use GPIOTE channels. */
+			nrf_gpiote_polarity_t pol;
+
+			if (data->double_edge & BIT(pin)) {
+				pol = NRF_GPIOTE_POLARITY_TOGGLE;
+			} else if (((data->active_level & BIT(pin)) != 0)
+				   ^ ((BIT(pin) & data->inverted) != 0)) {
+				pol = NRF_GPIOTE_POLARITY_LOTOHI;
+			} else {
+				pol = NRF_GPIOTE_POLARITY_HITOLO;
+			}
+
+			res = gpiote_channel_alloc(abs_pin, pol);
```

Comment:
There is a bug here!
gpiote_channel_alloc returns the channel number so if more then one pin will enable edge trigger it will fail as channel will be non-zero!

Reply:
This must be fixed asap

Reply:
@carlescufi @Mierunski FYI

Reply:
besides it seems that when I disable callbacks they are still coming!
how was it tested?

Reply:
I disable the callback in the interrupt but I get another interrupt just after that....
This was working fine before this change

Reply:
https://github.com/zephyrproject-rtos/zephyr/issues/9535